pkg/velero/backup.go

```diff
@@ -0,0 +1,353 @@
+}
+
+// Exists checks whether the given backup exists.
+func (builder *BackupBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if backup %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.apiClient.VeleroV1().Backups(builder.Definition.Namespace).Get(
+		context.Background(), builder.Definition.Name, metav1.GetOptions{})
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// Create makes a backup according to the backup definition and stores the created object in the backup builder.
+func (builder *BackupBuilder) Create() (*BackupBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Creating backup %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	if !builder.Exists() {
+		builder.Object, err = builder.apiClient.VeleroV1().Backups(builder.Definition.Namespace).Create(
+			context.TODO(), builder.Definition, metav1.CreateOptions{})
+	}
+
+	return builder, err
+}
+
+// Update renovates the existing backup object with the backup definition in builder.
+func (builder *BackupBuilder) Update() (*BackupBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Updating backup %s in namespace %s", builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.apiClient.VeleroV1().Backups(builder.Definition.Namespace).Update(
+		context.TODO(), builder.Definition, metav1.UpdateOptions{})
+
+	return builder, err
+}
+
+// Delete removes the backup object and resets the builder object.
+func (builder *BackupBuilder) Delete() (*BackupBuilder, error) {
+	if valid, err := builder.validate(); !valid {
+		return builder, err
+	}
+
+	glog.V(100).Infof("Deleting backup %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	if !builder.Exists() {
+		return builder, fmt.Errorf("backup cannot be deleted because it does not exist")
+	}
+
+	err := builder.apiClient.VeleroV1().Backups(builder.Definition.Namespace).Delete(
+		context.TODO(), builder.Object.Name, metav1.DeleteOptions{})
+
+	if err != nil {
+		return builder, fmt.Errorf("can not delete backup: %w", err)
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// validate will check that the builder and builder definition are properly initialized before
+// accessing any member fields.
+func (builder *BackupBuilder) validate() (bool, error) {
+	resourceCRD := "Backup"
+
+	if builder == nil {
+		glog.V(100).Infof("The %s builder is uninitialized", resourceCRD)
+
+		return false, fmt.Errorf("error: received nil %s builder", resourceCRD)
+	}
+
+	if builder.Definition == nil {
+		glog.V(100).Infof("The %s is undefined", resourceCRD)
+
+		builder.errorMsg = msg.UndefinedCrdObjectErrString(resourceCRD)
```

Comment:
I think in general, not necessarily this PR, but anytime an `errorMsg` is set, we should have an immediate `return`.  Doesn't have to change for this PR but maybe for a future effort.

Reply:
I don't have the power to resolve these comments once they are created so you might have to close these for me.